### ISSUE TITLE
problem with gradient check

### ISSUE THREAD

**Issue by glample** [Element-Research/rnn#203]
Hi,

I'm having an issue when I try to check the gradient of the following code:

```lua
require('torch')
require('nn')
require('nngraph')
require('optim')
require('rnn')
nn.FastLSTM.usenngraph = true

nWords = 6
nActions = 3
wordEmbDim = 4
lstmHidDim = 4

seq = nn.Sequencer(
    nn.Sequential()
        :add(nn.LookupTable(nWords, wordEmbDim))
        :add(nn.FastLSTM(wordEmbDim, lstmHidDim))
        :add(nn.Linear(wordEmbDim, nActions))
        :add(nn.LogSoftMax())
)

criterion = nn.SequencerCriterion(nn.CrossEntropyCriterion())

testInput = {torch.IntTensor{3, 2}, torch.IntTensor{3, 4}}
testOutput = {torch.IntTensor{1, 3}, torch.IntTensor{2, 3}}

parameters, grads = seq:getParameters()

function f(x)
    parameters:copy(x)
    -- seq:forget()
    seq:zeroGradParameters()
    seq:forward(testInput)
    criterion:forward(seq.output, testOutput)
    seq:backward(testInput, criterion:backward(seq.output, testOutput))
    return criterion.output, grads
end

err = optim.checkgrad(f, parameters:clone())
print(err)

```

The error is pretty big (about 0.01), but if I comment the line ":add(nn.FastLSTM(wordEmbDim, lstmHidDim))" the error is below 1e-6 and is a priori correct.
Am I doing something wrong when I add the LSTM?

Thanks!

**Comment by nicholas-leonard** [comment]
LookupTable causes this kind of thing. Not sure why though. Anyway, removing it from the test will generate small errors (1e-6, 1e-7): 
```lua
require('torch')
require('nn')
require('nngraph')
require('optim')
require('rnn')
nn.FastLSTM.usenngraph = true

nWords = 6
nActions = 3
wordEmbDim = 4
lstmHidDim = 4

seq = nn.Sequencer(
    nn.Sequential()
        --:add(nn.LookupTable(nWords, wordEmbDim))
        :add(nn.FastLSTM(wordEmbDim, lstmHidDim))
        :add(nn.Linear(lstmHidDim, nActions))
        :add(nn.LogSoftMax())
)

criterion = nn.SequencerCriterion(nn.ClassNLLCriterion())

--testInput = {torch.IntTensor{3, 2}, torch.IntTensor{3, 4}}
testInput = {torch.randn(2, wordEmbDim), torch.randn(2, wordEmbDim)}
testOutput = {torch.IntTensor{1, 3}, torch.IntTensor{2, 3}}

parameters, grads = seq:getParameters()

function f(x)
    parameters:copy(x)
    -- seq:forget()
    seq:zeroGradParameters()
    seq:forward(testInput)
    criterion:forward(seq.output, testOutput)
    seq:backward(testInput, criterion:backward(seq.output, testOutput))
    return criterion.output, grads
end

err = optim.checkgrad(f, parameters:clone())
print(err)
```

**Comment by glample** [comment]
Replacing the Lookuptable by a Linear also fails. Basically it seems to fail when there is something before the LSTM in the sequential.

**Comment by nicholas-leonard** [comment]
@glample You were right. There was a bug. Fixed in https://github.com/Element-Research/rnn/commit/e8dc44d42fcbabcb3d5e7bc24e63fcabe1ab7dbf . Thank you for raising the issue (even after I closed it) :)
